libraries/AP_Mount/AP_Mount_Topotek.cpp

```diff
@@ -0,0 +1,995 @@
+
+    _set_roll_angle_cmd[14] = hex2char((speed >> 4) & 0x0F);
+    _set_roll_angle_cmd[15] = hex2char((speed) & 0x0F);
+
+    send_packet(_set_roll_angle_cmd, 16);
+}
+
+// set rate target in radians/s
+void AP_Mount_Topotek::send_rate_target(float roll_rads, float pitch_rads, float yaw_rads, bool yaw_is_ef)
+{
+    // set gimbal's lock state if it has changed
+    if (!set_gimbal_lock(yaw_is_ef)) {
+        return;
+    }
+
+    // convert and constrain rates
+    const uint8_t roll_angle_speed = constrain_int16(degrees(roll_rads) * ANGULAR_VELOCITY_CONVERSION, -99, 99);
+    const uint8_t pitch_angle_speed = constrain_int16(degrees(pitch_rads) * ANGULAR_VELOCITY_CONVERSION, -99, 99);
+    const uint8_t yaw_angle_speed = constrain_int16(degrees(yaw_rads) * ANGULAR_VELOCITY_CONVERSION, -99, 99);
+
+    // send stop rotation command three times if target roll, pitch and yaw are zero
+    if (roll_angle_speed == 0 && pitch_angle_speed == 0 && yaw_angle_speed == 0) {
+        if (_stop_order_count < 3) {
+            _stop_order_count++;
+            send_packet(_gimbal_control_stop, 12);
+        }
+        return;
+    }
+    _stop_order_count = 0;
+
+    // prepare and send command
+    _set_yaw_pitch_roll_speed_cmd[10] = hex2char((yaw_angle_speed >> 4) & 0x0F);
+    _set_yaw_pitch_roll_speed_cmd[11] = hex2char((yaw_angle_speed) & 0x0F);
+
+    _set_yaw_pitch_roll_speed_cmd[12] = hex2char((pitch_angle_speed >> 4) & 0x0F);
+    _set_yaw_pitch_roll_speed_cmd[13] = hex2char((pitch_angle_speed) & 0x0F);
+
+    _set_yaw_pitch_roll_speed_cmd[14] = hex2char((roll_angle_speed >> 4) & 0x0F);
+    _set_yaw_pitch_roll_speed_cmd[15] = hex2char((roll_angle_speed) & 0x0F);
+
+    send_packet(_set_yaw_pitch_roll_speed_cmd, 16);
+}
+
+// send time and date to gimbal
+bool AP_Mount_Topotek::send_time_to_gimbal()
+{
+#if AP_RTC_ENABLED
+    // get date and time
+    // year is the regular Gregorian year, month is 0~11, day is 1~31, hour is 0~23, minute is 0~59, second is 0~60 (1 leap second), ms is 0~999
+    uint16_t year, ms;
+    uint8_t month, day, hour, min, sec;
+    if (!AP::rtc().get_date_and_time_utc(year, month, day, hour, min, sec, ms)) {
+        return false;
+    }
+
+    sprintf((char*)_set_gimbal_time + 10, "%02d%02d%02d%02d%02d%02d",
+            hour,
+            min,
+            sec,
+            day,
+            month + 1,
+            year - 2000);
+
+    return send_packet(_set_gimbal_time, 22);
+#else
+    return false;
+#endif
+}
+
+// send GPS-related information to the gimbal
+bool AP_Mount_Topotek::send_location_info(void)
+{
+    // get current location
+    Location loc;
+    int32_t alt_amsl_cm = 0;
+    if (!AP::ahrs().get_location(loc) || !loc.get_alt_cm(Location::AltFrame::ABSOLUTE, alt_amsl_cm)) {
+        return false;
+    }
+    int32_t lat = loc.lat;
+    int32_t lng = loc.lng;
+
+    double latitude = lat / 10000000.0;
+    double longitude = lng / 10000000.0;
```

Comment:
Don't need to have those

Reply:
Txs, I've made send_location_info() more concise although the two lines you mention are still mostly there.  please tell me if you're not OK with the new form.